src/nnet3/nnet-simple-component.cc

```diff
@@ -295,6 +295,96 @@ void NoOpComponent::Backprop(const std::string &debug_info,
+}
+
+void GradientClipperComponent::Write(std::ostream &os, bool binary) const {
+  WriteToken(os, binary, "<GradientClipperComponent>");
+  WriteToken(os, binary, "<Dim>");
+  WriteBasicType(os, binary, dim_);
+  WriteToken(os, binary, "<ClippingThreshold>");
+  WriteBasicType(os, binary, clipping_threshold_);
+  WriteToken(os, binary, "<NormBasedClipping>");
+  WriteBasicType(os, binary, norm_based_clipping_);
+  WriteToken(os, binary, "</GradientClipperComponent>");
+}
+
+std::string GradientClipperComponent::Info() const {
+  std::stringstream stream;
+  stream << Type() << ", dim=" << dim_
+         << ", norm-based-clipping=" << norm_based_clipping_
+         << ", clipping-threshold=" << clipping_threshold_;
+  return stream.str();
+}
+
+void GradientClipperComponent::Init(int32 dim, 
+                                    BaseFloat clipping_threshold,
+                                    bool norm_based_clipping)  {
+  KALDI_ASSERT(clipping_threshold >= 0 && dim > 0);
+  dim_ = dim;
+  clipping_threshold_ = clipping_threshold;
+  norm_based_clipping_ = norm_based_clipping;
+
+}
+
+void GradientClipperComponent::InitFromConfig(ConfigLine *cfl) {
+  int32 dim = 0;
+  bool ok = cfl->GetValue("dim", &dim);
+  bool norm_based_clipping = false;
+  BaseFloat clipping_threshold = 15.0;
+  cfl->GetValue("clipping-threshold", &clipping_threshold);
+  cfl->GetValue("norm-based-clipping", &norm_based_clipping);
+  if (!ok || cfl->HasUnusedValues() ||
+      clipping_threshold < 0 || dim <= 0)
+    KALDI_ERR << "Invalid initializer for layer of type "
+              << Type() << ": \"" << cfl->WholeLine() << "\"";
+  Init(dim, clipping_threshold, norm_based_clipping);
+}
+
+void GradientClipperComponent::Propagate(
+                                 const ComponentPrecomputedIndexes *indexes,
+                                 const CuMatrixBase<BaseFloat> &in,
+                                 CuMatrixBase<BaseFloat> *out) const {
+  out->CopyFromMat(in);
+}
+
+void GradientClipperComponent::Backprop(const std::string &debug_info,
+                             const ComponentPrecomputedIndexes *indexes,
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &out_deriv,
+                             Component *to_update, // may be NULL; may be identical
+                             // to "this" or different.
+                             CuMatrixBase<BaseFloat> *in_deriv) const {
+  in_deriv->CopyFromMat(out_deriv);
+  if (clipping_threshold_ > 0) {
+    if (norm_based_clipping_) {
+      BaseFloat deriv_norm = in_deriv->FrobeniusNorm();
+      in_deriv->Scale(clipping_threshold_ / deriv_norm);
+    } else {
+      in_deriv->ApplyCeiling(clipping_threshold_);
+      in_deriv->ApplyFloor(-1 * clipping_threshold_);
+    }
+    if (GetVerboseLevel() >= 2) {
+      BaseFloat norm_before_clipping = out_deriv.FrobeniusNorm(),
```

Comment:
this code would have to be changed if you allow backprop in place (but anyway you are redundantly computing the FrobeniusNorma()).